File: typescript/src/swagger20-schema.ts
import {
  isSwagger20Object,
  stringMember,
  type Swagger20Object,
  type Swagger20ReferenceGraphContract,
  type Swagger20Resource,
} from "./swagger20-model.js";
import { newSwagger20ResolutionMemo } from "./swagger20-reference.js";

/** Binding-relevant resolved declaration facts; no 3.x Schema type is reused. */
export interface Swagger20SchemaDeclaration {
  typed: boolean;
  types: Set<string>;
  formats: Set<string>;
}

export async function resolveSwagger20SchemaDeclaration(
  graph: Swagger20ReferenceGraphContract,
  value: unknown,
  resource: Swagger20Resource,
  allowFile = false,
  active = new Set<string>(),
): Promise<Swagger20SchemaDeclaration> {
  if (!isSwagger20Object(value)) throw new Error("Schema Object is not an object");
  const reference = stringMember(value, "$ref");
  if (reference.present) {
    if (!reference.valid || reference.value === "") throw new Error("Schema Object has an invalid $ref");
    const key = `${resource.retrieval ?? resource.requested ?? ""}|schema|${reference.value}`;
    if (active.has(key)) throw new Error("selected Schema reference cycle does not resolve a carriage declaration");
    active.add(key);
    try {
      const resolved = await graph.resolveReference(reference.value!, resource, newSwagger20ResolutionMemo());
      if (resolved.cycle) throw new Error("selected Schema reference cycle does not resolve a carriage declaration");
      return resolveSwagger20SchemaDeclaration(graph, resolved.node, resolved.resource, allowFile, active);
    } finally {
      active.delete(key);
    }
  }
  let declaration: Swagger20SchemaDeclaration = { typed: false, types: new Set(), formats: new Set() };
  if (Object.hasOwn(value, "type")) {
    declaration.typed = true;
    declaration.types = schemaTypes(value.type, allowFile);
  }
  const format = stringMember(value, "format");
  if (format.present) {
    if (!format.valid) throw new Error("Schema Object format is not a string");
    declaration.formats.add(format.value!);
  }
  if (Object.hasOwn(value, "allOf")) {
    if (!Array.isArray(value.allOf) || value.allOf.length === 0) throw new Error("Schema Object allOf is not a nonempty array");
    for (let index = 0; index < value.allOf.length; index++) {
      try {
        declaration = conjoin(declaration, await resolveSwagger20SchemaDeclaration(
          graph, value.allOf[index], resource, allowFile, active,
        ));
      } catch (error: unknown) {
        throw new Error(`Schema Object allOf member ${index}: ${errorMessage(error)}`, { cause: error });
      }
    }
  }
  return declaration;
}

function schemaTypes(value: unknown, allowFile: boolean): Set<string> {
  const values = Array.isArray(value) ? value : [value];
  if (values.length === 0) throw new Error("Schema Object type array is empty");
  const result = new Set<string>();
  for (const member of values) {
    if (typeof member !== "string" || member === "") throw new Error("Schema Object type contains a non-string or empty member");
    if (!["null", "boolean", "object", "array", "number", "string", "integer"].includes(member)
      && !(allowFile && member === "file")) throw new Error(`Schema Object type ${JSON.stringify(member)} is not admitted`);
    if (result.has(member)) throw new Error(`Schema Object type repeats ${JSON.stringify(member)}`);
    result.add(member);
  }
  return result;
}

function conjoin(left: Swagger20SchemaDeclaration, right: Swagger20SchemaDeclaration): Swagger20SchemaDeclaration {
  const typed = left.typed || right.typed;
  let types = new Set<string>();
  if (left.typed && right.typed) types = new Set([...left.types].filter((type) => right.types.has(type)));
  else if (left.typed) types = new Set(left.types);
  else if (right.typed) types = new Set(right.types);
  return { typed, types, formats: new Set([...left.formats, ...right.formats]) };
}

export function swagger20SoleString(declaration: Swagger20SchemaDeclaration): boolean {
  return declaration.typed && declaration.types.has("string")
    && [...declaration.types].every((type) => type === "string" || type === "null");
}

export function swagger20RawOctets(declaration: Swagger20SchemaDeclaration): boolean {
  if (declaration.typed && declaration.types.size === 1 && declaration.types.has("file")) return true;
  return swagger20SoleString(declaration) && declaration.formats.size === 1 && declaration.formats.has("binary");
}

export function swagger20ByteString(declaration: Swagger20SchemaDeclaration): boolean {
  return swagger20SoleString(declaration) && declaration.formats.size === 1 && declaration.formats.has("byte");
}

export function swagger20SchemaObject(value: unknown): Swagger20Object {
  if (!isSwagger20Object(value)) throw new Error("Schema Object is not an object");
  return value;
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: go/swagger20_schema.go
package openapiclient

import (
	"fmt"
	"strings"
)

type swagger20SchemaDeclaration struct {
	typed   bool
	types   map[string]bool
	formats map[string]bool
}

func resolveSwagger20SchemaDeclaration(graph *swagger20ReferenceGraph, value any, resource *swagger20Resource, allowFile bool) (swagger20SchemaDeclaration, error) {
	return resolveSwagger20SchemaDeclarationActive(graph, value, resource, allowFile, map[string]bool{})
}

func resolveSwagger20SchemaDeclarationActive(graph *swagger20ReferenceGraph, value any, resource *swagger20Resource, allowFile bool, active map[string]bool) (swagger20SchemaDeclaration, error) {
	object, ok := value.(map[string]any)
	if !ok {
		return swagger20SchemaDeclaration{}, fmt.Errorf("Schema Object is not an object")
	}
	raw := swagger20Object(object)
	if reference, present := raw.member("$ref"); present {
		ref, valid := reference.(string)
		if !valid || ref == "" {
			return swagger20SchemaDeclaration{}, fmt.Errorf("Schema Object has an invalid $ref")
		}
		key := artifactResourceKey(resource.base()) + "|schema|" + ref
		if active[key] {
			return swagger20SchemaDeclaration{}, fmt.Errorf("selected Schema reference cycle does not resolve a carriage declaration")
		}
		active[key] = true
		defer delete(active, key)
		resolved, err := graph.resolveReference(ref, resource, newSwagger20ResolutionMemo())
		if err != nil {
			return swagger20SchemaDeclaration{}, err
		}
		if resolved.cycle {
			return swagger20SchemaDeclaration{}, fmt.Errorf("selected Schema reference cycle does not resolve a carriage declaration")
		}
		return resolveSwagger20SchemaDeclarationActive(graph, resolved.node, resolved.resource, allowFile, active)
	}

	declaration := swagger20SchemaDeclaration{}
	if typeValue, present := raw.member("type"); present {
		types, err := swagger20SchemaTypes(typeValue, allowFile)
		if err != nil {
			return swagger20SchemaDeclaration{}, err
		}
		declaration.typed = true
		declaration.types = types
	}
	if format := raw.string("format"); format.present {
		if !format.valid {
			return swagger20SchemaDeclaration{}, fmt.Errorf("Schema Object format is not a string")
		}
		declaration.formats = map[string]bool{format.value: true}
	}
	if branches := raw.array("allOf"); branches.present {
		if !branches.valid || len(branches.value) == 0 {
			return swagger20SchemaDeclaration{}, fmt.Errorf("Schema Object allOf is not a nonempty array")
		}
		for index, branch := range branches.value {
			resolved, err := resolveSwagger20SchemaDeclarationActive(graph, branch, resource, allowFile, active)
			if err != nil {
				return swagger20SchemaDeclaration{}, fmt.Errorf("Schema Object allOf member %d: %w", index, err)
			}
			declaration = conjoinSwagger20SchemaDeclarations(declaration, resolved)
		}
	}
	return declaration, nil
}

func swagger20SchemaTypes(value any, allowFile bool) (map[string]bool, error) {
	values := []any{value}
	if array, ok := value.([]any); ok {
		if len(array) == 0 {
			return nil, fmt.Errorf("Schema Object type array is empty")
		}
		values = array
	}
	result := make(map[string]bool, len(values))
	for _, member := range values {
		name, ok := member.(string)
		if !ok || name == "" {
			return nil, fmt.Errorf("Schema Object type contains a non-string or empty member")
		}
		switch name {
		case "null", "boolean", "object", "array", "number", "string", "integer":
		case "file":
			if !allowFile {
				return nil, fmt.Errorf("Schema Object type file is admitted only at a response root")
			}
		default:
			return nil, fmt.Errorf("Schema Object type %q is not admitted", name)
		}
		if result[name] {
			return nil, fmt.Errorf("Schema Object type repeats %q", name)
		}
		result[name] = true
	}
	return result, nil
}

func conjoinSwagger20SchemaDeclarations(left, right swagger20SchemaDeclaration) swagger20SchemaDeclaration {
	result := swagger20SchemaDeclaration{}
	switch {
	case left.typed && right.typed:
		result.typed = true
		result.types = map[string]bool{}
		for name := range left.types {
			if right.types[name] {
				result.types[name] = true
			}
		}
	case left.typed:
		result.typed = true
		result.types = cloneSwagger20StringSet(left.types)
	case right.typed:
		result.typed = true
		result.types = cloneSwagger20StringSet(right.types)
	}
	result.formats = cloneSwagger20StringSet(left.formats)
	if result.formats == nil && len(right.formats) > 0 {
		result.formats = map[string]bool{}
	}
	for format := range right.formats {
		result.formats[format] = true
	}
	return result
}

func cloneSwagger20StringSet(input map[string]bool) map[string]bool {
	if len(input) == 0 {
		return nil
	}
	result := make(map[string]bool, len(input))
	for name := range input {
		result[name] = true
	}
	return result
}

func (d swagger20SchemaDeclaration) admitsStringAsSoleNonNullType() bool {
	if !d.typed || !d.types["string"] {
		return false
	}
	for name := range d.types {
		if name != "string" && name != "null" {
			return false
		}
	}
	return true
}

func (d swagger20SchemaDeclaration) format() (string, bool) {
	if len(d.formats) == 0 {
		return "", false
	}
	if len(d.formats) != 1 {
		return "", true
	}
	for format := range d.formats {
		return format, false
	}
	return "", false
}

func (d swagger20SchemaDeclaration) rawOctets() bool {
	if d.typed && len(d.types) == 1 && d.types["file"] {
		return true
	}
	format, conflict := d.format()
	return !conflict && d.admitsStringAsSoleNonNullType() && format == "binary"
}

func (d swagger20SchemaDeclaration) artifactByteString() bool {
	format, conflict := d.format()
	return !conflict && d.admitsStringAsSoleNonNullType() && format == "byte"
}

func isSwagger20JSONMedia(base string) bool {
	if base == "application/json" {
		return true
	}
	_, subtype, ok := strings.Cut(base, "/")
	return ok && strings.HasSuffix(subtype, "+json")
}

func isSwagger20CharacterMedia(base string) bool {
	primary, subtype, ok := strings.Cut(base, "/")
	if !ok {
		return false
	}
	if primary == "text" {
		return true
	}
	return base == "application/xml" || strings.HasSuffix(subtype, "+xml")
}
